scanner/k8s-assets/processor/processor.go

```diff
@@ -0,0 +1,179 @@
+	var serviceId string
+
+	listServicesFilter := client.ServiceFilter{
+		ServiceName: []string{serviceInfo.Name},
+	}
+	listServicesResp, err := client.ListServices(ctx, *p.Client, &listServicesFilter)
+	if err != nil {
+		return "", fmt.Errorf("Couldn't list services")
+	}
+
+	// Return the first item
+	if listServicesResp.Services.TotalCount > 0 {
+		for _, s := range listServicesResp.Services.Edges {
+			serviceId = s.Node.Id
+			break
+		}
+	} else {
+		return "", fmt.Errorf("ListServices returned no ServiceID")
+	}
+
+	return serviceId, nil
+}
+
+// ProcessPod processes a single pod and its containers
+func (p *Processor) ProcessPod(ctx context.Context, namespace string, serviceID string, podInfo scanner.PodInfo) error {
+	for _, containerInfo := range podInfo.Containers {
+		if err := p.ProcessContainer(ctx, namespace, serviceID, podInfo, containerInfo); err != nil {
+			return fmt.Errorf("failed to process container %s in pod %s: %w", containerInfo.Name, podInfo.Name, err)
+		}
+	}
+	return nil
+}
+
+func (p *Processor) getComponentVersion(ctx context.Context, manifest string) (string, error) {
+	var componentVersionId string
+
+	listComponentVersionFilter := client.ComponentVersionFilter{
+		Version: []string{manifest},
+	}
+	listCompoVersResp, err := client.ListComponentVersions(ctx, *p.Client, &listComponentVersionFilter)
+	if err != nil {
+		return "", fmt.Errorf("Couldn't list ComponentVersion")
+	}
+
+	if listCompoVersResp.ComponentVersions.TotalCount > 0 {
+		for _, cv := range listCompoVersResp.ComponentVersions.Edges {
+			componentVersionId = cv.Node.Id
+			break
+		}
+	} else {
+		return "", fmt.Errorf("ListComponentVersion returned no ComponentVersion objects")
+	}
+	return componentVersionId, nil
+}
+
+// ProcessContainer creates a ComponentVersion and ComponentInstance for a container
+func (p *Processor) ProcessContainer(ctx context.Context, namespace string, serviceID string, podInfo scanner.PodInfo, containerInfo scanner.ContainerInfo) error {
+
+	// Find component version by container image hash
+	componentVersionId, err := p.getComponentVersion(ctx, containerInfo.ImageHash)
+	if err != nil {
+		return fmt.Errorf("Couldn't find ComponentVersion")
+	}
+
+	// Create new CCRN
+	ccrn := CCRN{
+		Region:    p.config.RegionName,
+		Domain:    "x", // TODO
+		Project:   "x", // TODO
+		Cluster:   p.config.ClusterName,
+		Namespace: namespace,
+		Pod:       podInfo.Name,
+		PodID:     podInfo.Name, // Change this!
+		Container: containerInfo.Name,
+	}
+
+	// Create new ComponentInstance
+	componentInstanceInput := &client.ComponentInstanceInput{
+		Ccrn:               ccrn.String(),
+		Count:              len(podInfo.Containers),
+		ComponentVersionId: componentVersionId,
+		ServiceId:          serviceID,
+	}
+	createCompInstResp, err := client.CreateComponentInstance(ctx, *p.Client, componentInstanceInput)
+	if err != nil {
+		return fmt.Errorf("failed to create ComponentInstance: %w", err)
+
+	}
```

Comment:
We creating here a componentInstance per container. 

The problem with this is that we want to use logical grouping to avoid individual tracking of single containers. This means we actually group ALL containers of a certain version within a certain generator group and simply increase the count. Right now, we would create, to my understanding, a ComponentInstance per pod  per container. 

We could argue to do this as well and have a grouped self-reference. But for the first iteration, it's probably more easy to just do the count increase.

Reply:
Implemented using PodReplicaSet.